solana-agave: init at 2.1.21

TomMD opened the pull request:
Add a derivation of the tooling at https://github.com/anza-xyz/agave. This is a solana network validator and cli tooling.

1. Add TomMD as a maintainer
2. Add solana-agave package

## Things done

- Built on platform(s)
  - [X] x86_64-linux
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)

drupol reviewed (changes requested):
Hello,

Thanks for your PR!

Question: have you read the contribution guide?

This project is so big that there are strict git rules when it comes to submitting changes. Along the year we adopted particular practices and naming to facilitate the reviews, it would be nice if you could have a look at them and update your PR.

Hint: have a look at 99% of the PRs in this project to get some inspiration.

I'll set your PR to draft mode, feel free to set it to ready for review when you think it is ready.

Thanks

TomMD committed:
maintainers: add TomMD

TomMD committed:
solana-agave: init at 2.3.12

Repo and this version's changelog: https://github.com/anza-xyz/agave/releases/tag/v2.3.12
The agave repo, named after the vaildator server, might be niche but the package includes the 'solana'
cli which is user-facing.

drupol reviewed (changes requested):
Have you had a look at the contributing guide to learn how to properly format the PR title and the commit message?

drupol commented:
Please update the PR title so that it matches the main commit message.

aikooo7 commented:
> CC @0xnetfox @happysalada and @aikooo7 .
> 
> You are maintainers of the "solana-cli" package, but that package is based on a source repository that is deprecated after having been folded into the agave repository (which is packaged here in this PR as solana-agave). This is merely an FYI and I'm not asking for anything but seems appropriate you should get a FYI ping.

If your way is the most appropriate way and the most up-to-date then for me is fine

drupol commented on pkgs/by-name/so/solana-agave/package.nix:
```diff
@@ -0,0 +1,113 @@
+  canRun = stdenv.hostPlatform.emulatorAvailable buildPackages;
+  agave = "${stdenv.hostPlatform.emulator buildPackages} $out/bin/agave";
+in
+rustPlatform.buildRustPackage rec {
```
I think we should use the `finalAttrs` pattern here.

The `finalAttrs` pattern will let you remove the `rec` keyword ([see its implementation in Nix](https://github.com/NixOS/nix/blob/2d1d81114d72ace89ce08cd3bc93f4eb27a2975d/src/libexpr/nixexpr.cc#L121)).

Why is this pattern preferred to `rec` ?

Let's take this simple code example:

```nix
mkDerivation rec {
   foo = 1;
   bar = foo + 1;
}
```

and then `.overrideAttrs(old: { foo = 2; })`, you'll get `{ foo = 2; bar = 2; }` while with `finalAttrs` pattern, it would work correctly because it's a real fixed point.

Let me share a couple of useful links regarding the `finalAttrs` pattern:

1. History: https://discourse.nixos.org/t/avoid-rec-expresions-in-nixpkgs/8293
2. Documentation: https://nixos.org/manual/nixpkgs/unstable/#mkderivation-recursive-attributes
3. Recent example of implementation: https://github.com/NixOS/nixpkgs/compare/17f96f7b978e61576cfe16136eb418f74fab9952..9e6ea843e473d34d4f379b8b0d8ef0425a06defe

Feel free to reach out if you need some assistance.

happysalada commented:
Hey, thank you for the info !
We should probably add a throw alias for those consuming solana-cli as agave doesnt look like a drop in replacement. Do you know how to do that ? If you search for throw there should be examples. Otherwise im happy to do that in a separate pr when i have a moment.

drupol commented on pkgs/by-name/so/solana-agave/package.nix:
```diff
@@ -0,0 +1,103 @@
+      libz
+      rustPlatform.bindgenHook
+    ]
+    ++ lib.optional stdenv.hostPlatform.isDarwin apple-sdk_11
```
```suggestion
    ++ lib.optionals stdenv.hostPlatform.isDarwin [ apple-sdk_11 ]
```

drupol commented on pkgs/by-name/so/solana-agave/package.nix:
```diff
@@ -0,0 +1,103 @@
+  # Almost certainly caused by the ledger-tool test calling `Command::cargo_bin` which assumes a good bit about the current environment.
+  doCheck = false;
+
+  meta = with lib; {
```
```suggestion
  meta = {
```

drupol commented on pkgs/by-name/so/solana-agave/package.nix:
```diff
@@ -0,0 +1,103 @@
+    description = "Solana Network Validator";
+    homepage = "https://github.com/anza-xyz/agave";
+    changelog = "https://github.com/anza-xyz/agave/releases/tag/${version}";
+    license = with licenses; [
```
```suggestion
    license = with lib.licenses; [
```

drupol commented on pkgs/by-name/so/solana-agave/package.nix:
```diff
@@ -0,0 +1,103 @@
+    license = with licenses; [
+      asl20
+    ];
+    maintainers = with maintainers; [
```
```suggestion
    maintainers = with lib.maintainers; [
```

drupol commented on pkgs/by-name/so/solana-agave/package.nix:
```diff
@@ -0,0 +1,103 @@
+    mainProgram = "agave";
+  };
+
+  # For the same reason as discussed in solana-cli derivation (crossbeam softlink), the no_atomic file is missing
+  # and either must somehow be rendered unneeded (using an upstream package) or replaced. A cleaner, non-behavior-changing,
+  # solution would be to commit the file to the repo fork (replacing the softlink).
+  cargoPatches = [
+    ./crossbeam-epoch.patch
+    ./qualifier_attr-dep.patch
+  ];
+
+  # If set, always finds OpenSSL in the system, even if the vendored feature is enabled.
+  OPENSSL_NO_VENDOR = 1;
```
The `meta` attribute is usually the last attribute in the file. Please move `cargoPatches` near the Cargo stuff up there.

The `OPENSSL_NO_VENDOR` should be under `env` attribute.

drupol commented on pkgs/by-name/so/solana-agave/package.nix:
```diff
@@ -0,0 +1,103 @@
+  llvmPackages,
+}:
+
+let
+  canRun = stdenv.hostPlatform.emulatorAvailable buildPackages;
+  agave = "${stdenv.hostPlatform.emulator buildPackages} $out/bin/agave";
+in
```
Those things can be removed, they are not used.

drupol commented on pkgs/by-name/so/solana-agave/package.nix:
```diff
@@ -0,0 +1,100 @@
+  src = fetchFromGitHub {
+    owner = "anza-xyz";
+    repo = "agave";
+    rev = "v${finalAttrs.version}";
```
Use tag

drupol commented on pkgs/by-name/so/solana-agave/package.nix:
```diff
@@ -0,0 +1,100 @@
+      libz
+      rustPlatform.bindgenHook
+    ]
+    ++ lib.optional stdenv.hostPlatform.isDarwin apple-sdk_11
```
```suggestion
    ++ lib.optionals stdenv.hostPlatform.isDarwin [ apple-sdk_11 ]
```

ttytm commented:
@TomMD we got a recent update in the major version of the solana-cli nixpkg.
Is there an outlook to finalize the agave PR? Thanks for the work that went into it already!
Best regards

drupol reviewed (approved):
Diff LGTM ! Can you please run nixpkgs-review with --post-result flag on this PR?

TomMD commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 394782`
Commit: `51ccfe27916ae8f43731a5b2b1b0c4a8d9697b75`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>solana-agave</li>
  </ul>
</details>
